zebra/rt_netlink.c

```diff
@@ -920,6 +930,15 @@ int netlink_route_read(struct zebra_ns *zns)
 	if (ret < 0)
 		return ret;
 
+	/*
+	 * Start a timer of 15 secs to clean stale kernel route after
+	 * reconciliation. Note netlink_route_read function is called
+	 * only at startup, so timer will be activated only once.
+	 */
+	if (kernel_reconcile)
+		thread_add_timer(zrouter.master, rib_sweep_kernel_reconcile_route,
```

Comment:
I did this analysis for BGP which is used heavily in DCs.  Below is the analysis without my fix.

------------------------
1555633942 = Friday, April 19, 2019 00:32:22 ( This is when zebra starts to sweep routes from kernel during startup, so kernel is blind after this point.)

--------------------------------------------------
(This is where Zebra started adding routes in kernel after BGP updates.)

Apr 19 00:32:24.747571 falco-test-dut01 NOTICE bgp#zebra[86]: frr-kernel: Added 1 routes. Del time 1555633942
Apr 19 00:32:24.876074 falco-test-dut01 NOTICE bgp#zebra[86]: frr-kernel: Added 1001 routes. Del time 1555633942
Apr 19 00:32:25.055259 falco-test-dut01 NOTICE bgp#zebra[86]: frr-kernel: Added 2001 routes. Del time 1555633942
Apr 19 00:32:25.210264 falco-test-dut01 NOTICE bgp#zebra[86]: frr-kernel: Added 3001 routes. Del time 1555633942
Apr 19 00:32:25.388983 falco-test-dut01 NOTICE bgp#zebra[86]: frr-kernel: Added 4001 routes. Del time 1555633942
Apr 19 00:32:25.527489 falco-test-dut01 NOTICE bgp#zebra[86]: frr-kernel: Added 5001 routes. Del time 1555633942
Apr 19 00:32:25.653258 falco-test-dut01 NOTICE bgp#zebra[86]: frr-kernel: Added 6001 routes. Del time 1555633942
Apr 19 00:32:25.789716 falco-test-dut01 NOTICE bgp#zebra[86]: frr-kernel: Added 7001 routes. Del time 1555633942
Apr 19 00:32:25.951092 falco-test-dut01 NOTICE bgp#zebra[86]: frr-kernel: Added 8001 routes. Del time 1555633942
Apr 19 00:32:28.949341 falco-test-dut01 NOTICE bgp#zebra[86]: frr-kernel: Added 9001 routes. Del time 1555633942
Apr 19 00:32:29.072326 falco-test-dut01 NOTICE bgp#zebra[86]: frr-kernel: Added 10001 routes. Del time 1555633942
---------------------------------------------------
It ~7 secs between first kernel route deletion and to add 10001 routes.

Reply:
For BGP, we were planning to pass EOR down to Zebra. But this fix is applicable for all protocols and there is no fix timer, which works for all protocols supported by Zebra.  So it is more or less expert`s advice which I will follow, to start with, it is 15 secs.

As far as my advice goes, I planned to add an argument with -K option, which will be timer. So while restarting user can choose the timer as per protocols, one is running.  Thanks.

Reply:
I think having timer value as an argument is good as user's can specify the timer values depending on their network topology and use cases.

Reply:
Yeah, I agree to that, because no fixed timer value is perfect for all protocols. I will add a argument with -K option. Thanks.

Reply:
The timer start does not belong in zebra/rt_netlink.c It should be main or zebra_router_init()

Reply:
I want to start this timer once startup route learning from kernel is done. i.e. zebra_ns_init()-> zebra_ns_enable()-> route_read()-> netlink_route_read().

So I can start the timer in main() soon after zebra_ns_init() call. Which seems good to me because both kernel_reconcile and timer variable will be available there and it is more visible way to call timer. 

Other option is inside zebra_ns_init(). I will prefer 1st one i.e. main().  Thanks.

Reply:
Calling timer in main now.